annotated allocator snapshots

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82146

Record stack trace information for each allocated segment in the allocator.
It takes around 1.5us to record 50 stack frames of context.
Since invoking a Pytorch operator is around 8us, this adds minimal overhead but we still leave it disabled by default so that we can test it more on real workloads first.

Stack information is kept both for allocated blocks and the last allocation used inactive blocks. We could potential keep around the _first_ allocation that caused the block to get allocated from cuda as well.

Potential Followups:
* stack frame entries are small (16 bytes), but the list of Frames is not compressed eventhough most frames will share some entries. So far this doesn't produce huge dumps (7MB for one real workload that uses all memory on the GPU), but it can be much smaller through compression.
* Code to format the information is slow (a few seconds) because it uses python and FlameGraph.pl
* Things allocated during the backward pass have no stack frames because they are run on another C++ thread.

zdevito committed:
[WIP] annotated allocator snapshots

Sketch of how we can record stack trace information for each allocated segment in the allocator.
It takes around 1.5us to record 50 stack frames of context.
Since invoking a Pytorch operator is around 8us, this adds minimal overhead.
Stack information is kept both for allocated blocks and the last allocation used inactive blocks.

TODO:
* stack frame entries are small (16 bytes), but the list of Frames is not compressed eventhough most frames will share some entries.
* Code to format the information is slow, so if there are a lot of allocations it will have performance problems.
* Things allocated during the backward pass have no stack frames because they are run on another C++ thread.

[ghstack-poisoned]

zdevito committed:
Update on "[WIP] annotated allocator snapshots"


Sketch of how we can record stack trace information for each allocated segment in the allocator.
It takes around 1.5us to record 50 stack frames of context.
Since invoking a Pytorch operator is around 8us, this adds minimal overhead.
Stack information is kept both for allocated blocks and the last allocation used inactive blocks.

TODO:
* stack frame entries are small (16 bytes), but the list of Frames is not compressed eventhough most frames will share some entries.
* Code to format the information is slow, so if there are a lot of allocations it will have performance problems.
* Things allocated during the backward pass have no stack frames because they are run on another C++ thread.

[ghstack-poisoned]

zdevito committed:
Update on "[WIP] annotated allocator snapshots"


Sketch of how we can record stack trace information for each allocated segment in the allocator.
It takes around 1.5us to record 50 stack frames of context.
Since invoking a Pytorch operator is around 8us, this adds minimal overhead.
Stack information is kept both for allocated blocks and the last allocation used inactive blocks.

TODO:
* stack frame entries are small (16 bytes), but the list of Frames is not compressed eventhough most frames will share some entries.
* Code to format the information is slow, so if there are a lot of allocations it will have performance problems.
* Things allocated during the backward pass have no stack frames because they are run on another C++ thread.

[ghstack-poisoned]

zdevito committed:
Update on "[WIP] annotated allocator snapshots"


Sketch of how we can record stack trace information for each allocated segment in the allocator.
It takes around 1.5us to record 50 stack frames of context.
Since invoking a Pytorch operator is around 8us, this adds minimal overhead.
Stack information is kept both for allocated blocks and the last allocation used inactive blocks.

TODO:
* stack frame entries are small (16 bytes), but the list of Frames is not compressed eventhough most frames will share some entries.
* Code to format the information is slow, so if there are a lot of allocations it will have performance problems.
* Things allocated during the backward pass have no stack frames because they are run on another C++ thread.

[ghstack-poisoned]

zdevito committed:
Update on "[WIP] annotated allocator snapshots"


Sketch of how we can record stack trace information for each allocated segment in the allocator.
It takes around 1.5us to record 50 stack frames of context.
Since invoking a Pytorch operator is around 8us, this adds minimal overhead.
Stack information is kept both for allocated blocks and the last allocation used inactive blocks.

TODO:
* stack frame entries are small (16 bytes), but the list of Frames is not compressed eventhough most frames will share some entries.
* Code to format the information is slow, so if there are a lot of allocations it will have performance problems.
* Things allocated during the backward pass have no stack frames because they are run on another C++ thread.

[ghstack-poisoned]

zdevito committed:
Update on "[WIP] annotated allocator snapshots"


Sketch of how we can record stack trace information for each allocated segment in the allocator.
It takes around 1.5us to record 50 stack frames of context.
Since invoking a Pytorch operator is around 8us, this adds minimal overhead.
Stack information is kept both for allocated blocks and the last allocation used inactive blocks.

TODO:
* stack frame entries are small (16 bytes), but the list of Frames is not compressed eventhough most frames will share some entries.
* Code to format the information is slow, so if there are a lot of allocations it will have performance problems.
* Things allocated during the backward pass have no stack frames because they are run on another C++ thread.

[ghstack-poisoned]

zdevito committed:
Update on "annotated allocator snapshots"


Record stack trace information for each allocated segment in the allocator.
It takes around 1.5us to record 50 stack frames of context.
Since invoking a Pytorch operator is around 8us, this adds minimal overhead but we still leave it disabled by default so that we can test it more on real workloads first.

Stack information is kept both for allocated blocks and the last allocation used inactive blocks. We could potential keep around the _first_ allocation that caused the block to get allocated from cuda as well.

Potential Followups:
* stack frame entries are small (16 bytes), but the list of Frames is not compressed eventhough most frames will share some entries. So far this doesn't produce huge dumps (7MB for one real workload that uses all memory on the GPU), but it can be much smaller through compression.
* Code to format the information is slow (a few seconds) because it uses python and FlameGraph.pl
* Things allocated during the backward pass have no stack frames because they are run on another C++ thread.

[ghstack-poisoned]

zdevito commented:
@pytorchbot rebase

albanD commented on torch/cuda/memory.py:
```diff
@@ -590,3 +594,23 @@ def mem_get_info(device: Union[Device, int] = None) -> Tuple[int, int]:
     device = _get_device_index(device)
     return torch.cuda.cudart().cudaMemGetInfo(device)
+
+def enable_memory_history(device: Union[Device, int] = None):
```
All these public functions are missing doc.

albanD replied:
This can't be disabled?

albanD commented on torch/cuda/memory_viz.py:
```diff
@@ -0,0 +1,188 @@
+    import os.path
+    thedir = os.path.realpath(os.path.dirname(__file__))
+    if thedir in sys.path:
+        # otherwise we find cuda/random.py as random...
```
Where is that a problem?

zdevito replied:
when you run this script right out of the repo. I use it on my Mac without even building PyTorch. We'll organize this better after figuring out what the final toolset will look like.

albanD replied:
Ok

albanD commented on torch/cuda/memory_viz.py:
```diff
@@ -0,0 +1,188 @@
+
+    fn_name = 'torch.cuda.memory_dbg.snapshot()'
+    pickled = f'pickled memory statistics from {fn_name}'
+    parser = argparse.ArgumentParser(description='Visualize memory dumps produced by {fn_name}')
```
missing the f for the fstring

albanD commented on torch/cuda/memory_viz.py:
```diff
@@ -0,0 +1,188 @@
+import io
+import subprocess
+from typing import Dict, Any
+__all__ = ["format_flamegraph", "segments", "memory", "compare", "stats", "Bytes"]
```
Most of these don't need to be public right?
If they are, they need doc :)

albanD commented on test/test_cuda.py:
```diff
@@ -4402,5 +4402,25 @@ class TestNamedTupleInput_1(NamedTuple):
             self.assertTrue(torch.equal(x, cat))
 
+    def test_memory_snapshot(self):
+        torch.cuda.memory.enable_memory_history()
```
Isn't this going to leak state to all the other tests?

albanD commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -1107,19 +1148,35 @@ class DeviceCachingAllocator {
+        src->history_last->next = std::move(dst->history);
+        dst->history = std::move(src->history);
+      }
+      src->history_last = nullptr;
```
Why reset only this one field?

zdevito replied:
`src->history` is already moved/cleared above if it was set to something

albanD replied:
Ho right the move, ok!

albanD commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -706,6 +735,17 @@ class DeviceCachingAllocator {
     block->allocated = true;
+    if (context) {
+      trimHistoryBefore(block, (char*)block->ptr + size);
+      block->history = std::unique_ptr<History>(new History{
```
why not make_unique?

albanD commented on torch/csrc/cuda/Module.cpp:
```diff
@@ -519,32 +519,109 @@ PyObject* THCPModule_resetPeakMemoryStats(PyObject* _unused, PyObject* arg) {
+    PyThreadState* tstate = PyThreadState_GET();
+    PyFrameObject* f = tstate->frame;
+    while (f) {
+      Py_XINCREF((PyObject*)f->f_code);
```
This is going to break 3.11 build.
Please use the correct getters from `python_compat.h` please.

albanD commented on torch/csrc/cuda/Module.cpp:
```diff
@@ -519,32 +519,109 @@ PyObject* THCPModule_resetPeakMemoryStats(PyObject* _unused, PyObject* arg) {
+    py::gil_scoped_acquire acquire;
+    auto r = std::make_unique<StackContext>();
+    PyThreadState* tstate = PyThreadState_GET();
+    PyFrameObject* f = tstate->frame;
```
nit; PyEval_GetFrame

albanD commented on torch/csrc/cuda/Module.cpp:
```diff
@@ -519,32 +519,109 @@ PyObject* THCPModule_resetPeakMemoryStats(PyObject* _unused, PyObject* arg) {
+            for (auto& f : sc->frames) {
+              py::dict frame;
+              frame[filename_s] =
+                  py::reinterpret_borrow<py::object>(f.code->co_filename);
```
Same here, use the corresponding getters

albanD commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -542,10 +557,19 @@ class DeviceCachingAllocator {
   }
 
+  void setContextRecorder(CreateContextFn c) {
+    context_recorder_ = std::move(c);
```
Is this thread safe?

albanD reviewed (commented):
Sounds mostly good.
What is the impact in perf when this is not used?

albanD commented:
Added binary builds to get python 3.11 build signal.

zdevito committed:
Update on "annotated allocator snapshots"


Record stack trace information for each allocated segment in the allocator.
It takes around 1.5us to record 50 stack frames of context.
Since invoking a Pytorch operator is around 8us, this adds minimal overhead but we still leave it disabled by default so that we can test it more on real workloads first.

Stack information is kept both for allocated blocks and the last allocation used inactive blocks. We could potential keep around the _first_ allocation that caused the block to get allocated from cuda as well.

Potential Followups:
* stack frame entries are small (16 bytes), but the list of Frames is not compressed eventhough most frames will share some entries. So far this doesn't produce huge dumps (7MB for one real workload that uses all memory on the GPU), but it can be much smaller through compression.
* Code to format the information is slow (a few seconds) because it uses python and FlameGraph.pl
* Things allocated during the backward pass have no stack frames because they are run on another C++ thread.

[ghstack-poisoned]

zdevito commented:
Update to:
* add option to disable history(old history gathered will still remain in the allocator, but it won't continue to record)
* use a std::atomic<funcptr> for the context gatherer so it can be thread safe while still being outside of the lock
* timed without history recording enabled, I cannot measure any difference in performance. This is probably because it only does 1 atomic load and 2 if statements of extra work in the allocator.
* updated to use the 3.11 APIs. I think co_filename and co_name were not among the things changed in 3.11

zdevito committed:
Update on "annotated allocator snapshots"


Record stack trace information for each allocated segment in the allocator.
It takes around 1.5us to record 50 stack frames of context.
Since invoking a Pytorch operator is around 8us, this adds minimal overhead but we still leave it disabled by default so that we can test it more on real workloads first.

Stack information is kept both for allocated blocks and the last allocation used inactive blocks. We could potential keep around the _first_ allocation that caused the block to get allocated from cuda as well.

Potential Followups:
* stack frame entries are small (16 bytes), but the list of Frames is not compressed eventhough most frames will share some entries. So far this doesn't produce huge dumps (7MB for one real workload that uses all memory on the GPU), but it can be much smaller through compression.
* Code to format the information is slow (a few seconds) because it uses python and FlameGraph.pl
* Things allocated during the backward pass have no stack frames because they are run on another C++ thread.

[ghstack-poisoned]

zdevito committed:
Update on "annotated allocator snapshots"


Record stack trace information for each allocated segment in the allocator.
It takes around 1.5us to record 50 stack frames of context.
Since invoking a Pytorch operator is around 8us, this adds minimal overhead but we still leave it disabled by default so that we can test it more on real workloads first.

Stack information is kept both for allocated blocks and the last allocation used inactive blocks. We could potential keep around the _first_ allocation that caused the block to get allocated from cuda as well.

Potential Followups:
* stack frame entries are small (16 bytes), but the list of Frames is not compressed eventhough most frames will share some entries. So far this doesn't produce huge dumps (7MB for one real workload that uses all memory on the GPU), but it can be much smaller through compression.
* Code to format the information is slow (a few seconds) because it uses python and FlameGraph.pl
* Things allocated during the backward pass have no stack frames because they are run on another C++ thread.

[ghstack-poisoned]

albanD commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -706,6 +737,17 @@ class DeviceCachingAllocator {
     block->allocated = true;
+    if (context) {
+      trimHistoryBefore(block, (char*)block->ptr + size);
+      block->history = std::make_unique<History>(History{
```
nit: Not need to create the temporary object here std::make_unique<History>(block->ptr, ...)

zdevito replied:
I thought so too but GCC thinks otherwise... something about {} vs () initialization...

albanD commented on c10/cuda/CUDACachingAllocator.h:
```diff
@@ -114,6 +130,7 @@ struct SegmentInfo {
   int64_t total_size = 0;
   int64_t allocated_size = 0;
   int64_t active_size = 0;
+  int64_t stream = 0;
```
nit: Why not cudaStream_t ?

zdevito replied:
I was just being consistent with how it will get exposed in Python, where I do not want to have to create a real stream object, because they can't be pickled. It's more of an ID for debugging than a handle back to the actual stream.

zdevito replied:
Since it doesn't matter in C++, I switched it so that the conversion to int happens in the Python bindings.

albanD commented on test/test_cuda.py:
```diff
@@ -4407,6 +4408,35 @@ class TestNamedTupleInput_1(NamedTuple):
+        try:
+            torch.cuda.memory.empty_cache()
+            torch.cuda.memory._record_memory_history(True)
+            x = torch.rand(311, 411).cuda()
```
nit: `torch.rand(311, 411, device="cuda")`

albanD commented on test/test_cuda.py:
```diff
@@ -4407,6 +4408,35 @@ class TestNamedTupleInput_1(NamedTuple):
+            tensors = [torch.rand(128, device='cuda') for _ in range(1000)]
+            while tensors:
+                del tensors[randint(0, len(tensors) - 1)]
+            torch.rand(128 * 5, device='cuda')
```
This is a really big Tensor for testing? Can we reduce all the sizes here or they are tuned to match the allocator block sizes? If so, can you leave a comment here to specific that?

zdevito replied:
It's the latter 128*sizeof(float) == 512, which is the block size these allocations would get rounded to otherwise. I'll comment.

albanD commented on torch/csrc/cuda/Module.cpp:
```diff
@@ -519,32 +519,110 @@ PyObject* THCPModule_resetPeakMemoryStats(PyObject* _unused, PyObject* arg) {
   using c10::cuda::CUDACachingAllocator::SegmentInfo;
 
-  const auto segmentInfoToDict = [](const SegmentInfo& segmentInfo) {
+  py::str device_s = "device";
```
nit you could have bound these to python in a private namespace to be able to access them from the python code

zdevito replied:
I wanted to avoid having to do static initialization of them. The previous version just passed strings to the pybind11 API, but they would end up creating a lot of string copies of the keys, which would then create large pickle archives when saving a memory dump. This was just the simplest fix for that.

albanD reviewed (approved):
A few nits but sounds good!

zdevito committed:
Update on "annotated allocator snapshots"


Record stack trace information for each allocated segment in the allocator.
It takes around 1.5us to record 50 stack frames of context.
Since invoking a Pytorch operator is around 8us, this adds minimal overhead but we still leave it disabled by default so that we can test it more on real workloads first.

Stack information is kept both for allocated blocks and the last allocation used inactive blocks. We could potential keep around the _first_ allocation that caused the block to get allocated from cuda as well.

Potential Followups:
* stack frame entries are small (16 bytes), but the list of Frames is not compressed eventhough most frames will share some entries. So far this doesn't produce huge dumps (7MB for one real workload that uses all memory on the GPU), but it can be much smaller through compression.
* Code to format the information is slow (a few seconds) because it uses python and FlameGraph.pl
* Things allocated during the backward pass have no stack frames because they are run on another C++ thread.

[ghstack-poisoned]

zdevito commented:
@pytorch merge

zdevito commented:
@pytorchbot merge

zdevito commented:
@pytorchbot merge -f "unrelated binary build failures that got added because we were checking Python 3.11 would compile"
